main/src/ca/uwaterloo/flix/api/lsp/provider/HighlightProvider.scala

```diff
@@ -261,58 +264,180 @@ object HighlightProvider {
+      case SymUse.TraitSymUse(sym, _) :: _ => Some(highlightTraitSym(sym))
+      case TraitConstraint.Head(sym, _) :: _ => Some(highlightTraitSym(sym))
       // Type Aliases
-      case TypedAst.TypeAlias(_, _, _, sym, _, _, _) => Some(highlightTypeAliasSym(sym))
-      case Type.Alias(AliasConstructor(sym, _), _, _, _) => Some(highlightTypeAliasSym(sym))
+      case TypedAst.TypeAlias(_, _, _, sym, _, _, _) :: _ => Some(highlightTypeAliasSym(sym))
+      case Type.Alias(AliasConstructor(sym, _), _, _, _) :: _ => Some(highlightTypeAliasSym(sym))
       // Type Variables
-      case TypedAst.TypeParam(_, sym, _) => Some(highlightTypeVarSym(sym))
-      case Type.Var(sym, _) => Some(highlightTypeVarSym(sym))
+      case TypedAst.TypeParam(_, sym, _) :: _ => Some(highlightTypeVarSym(sym))
+      case Type.Var(sym, _) :: _ => Some(highlightTypeVarSym(sym))
       // Variables
-      case Binder(sym, _) => Some(highlightVarSym(sym))
-      case TypedAst.Expr.Var(varSym, _, _) => Some(highlightVarSym(varSym))
+      case Binder(sym, _) :: _ => Some(highlightVarSym(sym))
+      case TypedAst.Expr.Var(varSym, _, _) :: _ => Some(highlightVarSym(varSym))
 
       case _ => None
     }
   }
 
+  private def highlightEffectTVars(scope: SourceLocation, tvar: Type.Var, loc: SourceLocation)(implicit root: Root, acceptor: Acceptor): JObject = {
+    var exps: List[SourceLocation] = Nil
+
+    def consider(exp: Expr): Unit = {
+      if (!exp.eff.typeVars.contains(tvar)) { return }
+      exps = exp.loc :: exps
+    }
+
+    val consumer = EffectConsumer(consider)
+
+    Visitor.visitRoot(root, consumer, acceptor)
+
+    val retEffHighlight = DocumentHighlight(Range.from(loc), DocumentHighlightKind.Write)
+    val expHighlights = exps.map(loc => DocumentHighlight(Range.from(loc), DocumentHighlightKind.Read))
+    val highlights = retEffHighlight :: expHighlights
+
+    ("status" -> ResponseStatus.Success) ~ ("result" -> JArray(highlights.map(_.toJSON)))
+
+  }
+
+  private def highlightEffectConcrete(scope: SourceLocation, sym: Symbol.EffectSym, loc: SourceLocation)(implicit root: Root, acceptor: Acceptor): JObject = {
+    var exps: List[SourceLocation] = Nil
+
+    def consider(exp: Expr): Unit = {
+      if (!exp.eff.effects.contains(sym)) { return }
+      // TODO also filter out anything not in `scope`
+      exps = exp.loc :: exps
+    }
+
+    val consumer = EffectConsumer(consider)
+
+    Visitor.visitRoot(root, consumer, acceptor)
+
+    val retEffHighlight = DocumentHighlight(Range.from(loc), DocumentHighlightKind.Write)
+    val expHighlights = exps.map(loc => DocumentHighlight(Range.from(loc), DocumentHighlightKind.Read))
+    val highlights = retEffHighlight :: expHighlights
+
+    ("status" -> ResponseStatus.Success) ~ ("result" -> JArray(highlights.map(_.toJSON)))
+  }
+
+  private case class EffectConsumer(consider: Expr => Unit) extends Consumer {
+    override def consumeExpr(exp: Expr): Unit = exp match {
+      case Expr.Cst(_, _, _) => ()
```

Comment:
This seems to defeat the entire point of using the Visitor?

Reply:
I agree that it's a lot, but it's just about identifying what kind of expression it is, so it's not really doing what the Visitors job. The reason I do it is so we can filter out the kinds of expression that we don't want to highlight. If we highlighted every Expr with the right effect, we'd highlight all the composite ones too, like `e1; e2` when we just care about `e2` etc.

Reply:
It could be make less verbose by introducing a catch-all case for all the kinds of expressions we don't care about. But if your concern is about separating concerns and responsibility then I suppose that doesn't solve anything

Reply:
Why not highlight just ApplyOp?

Reply:
There is no ApplyOp. There is `do` which does that I suppose, but it would just not cover all the sources of effects in a function body. For instance it wouldn't cover calling a function that has the effect in question. Then there are Array and Struct operations, effect casts and `do`. All of these are or may be impure. Then there's java interop stuff.

There are a couple of things on there that I could maybe remove. I don't believe  `Unary and Binary ops` can be impure. And then there are a couple of intrinsics that there is just no reason to support since they aren't allowed.

Reply:
What we could do is make the visitor gather all the expressions with the effect in question and then filter them afterwards. We would still need a big match case like that, but it would be separate from the consumer. Maybe that's better?

Reply:
@JonathanStarup also mentioned that that had been brought up as an idea for how to approach it

Reply:
we could do that, but then we only highlight do expressions. We want all sources of the effect to be highlighted, at least eventually, right?

Reply:
Do you just want a smaller pr initially?

Reply:
Doing the general "where does this effect come from" seems very hard (fx if `f(exp1, exp2)` has effect Crash, it could be because `f` introduces Crash, `f` could introduce crash because of a higher order effect of either exp1 or exp2, the application could have crash because exp1 or exp2 or both has Crash, in which case the application of f is not the cause of Crash). So even with this approach you can't get the full ideal solution. 

That why I think magnus just want the simple thing first - get it started with small steps

Reply:
oh, good point. I see. Alright. I'll just remove the other cases for now then

Reply:
Then we can maybe thing about a more sophisticated and full solution afterwards

Reply:
maybe some kind of consumer-acceptor combo that finds the precise origins of an effect

Reply:
I've limited the support to highlighting `do` expressions

Reply:
Sounds good - yeah i hope we get to it, seems complicated but also useful